src/java.base/share/classes/java/lang/reflect/code/type/CoreTypeFactory.java

```diff
@@ -114,6 +115,29 @@ public TypeElement constructType(TypeDefinition tree) {
                 }
                 typeArguments.add(a);
             }
+            if (identifier.equals("+") || identifier.equals("-")) {
+                // wildcard type
+                BoundKind kind = identifier.equals("+") ?
+                        BoundKind.EXTENDS : BoundKind.SUPER;
+                return JavaType.wildcard(kind, typeArguments.get(0));
+            } else if (identifier.contains("::")) {
+                // type-var
+                if (typeArguments.size() != 1) {
+                    throw new IllegalArgumentException("Bad type-variable bounds: " + tree);
+                }
+                String[] parts = identifier.split("::");
```

Comment:
And this is the duplicate parsing logic (although here we already know if it's a method or a class type-variable based on the number of `::`)

Reply:
I am wondering if instead we can check `#`, and the parsers job is dumbly accumulate all valid characters (selected tokens and identifiers) up to but not including the `<` token. We could even check if there is quoted string for the type identifier.

Note the special code for arrays in the parser was added only to avoid updating many tests.

Reply:
I can take a look

Reply:
Uploaded a new iteration with this simplification (which looks much nicer than what I had):

8ad6110bdb3c099b1421a04272c56685dd551f95

Note that if we wanted a truly general "quoting" mechanism we'd need both a prefix and a suffix token. Otherwise one can only use quotes if there's some nested type-definition with `<>`. Your idea of using just strings (e.g. surrounded with `"`) seems a powerful one (and more robust in the long run), because it would make the desc parsing logic a lot less opinionated (e.g. we wouldn't even need to special case qualified identifiers).

Reply:
That's much simpler. We can iterate further afterwards if need be. I believe you can now replace `identifier.contains("::")` with `identifier.startsWith("#")`?